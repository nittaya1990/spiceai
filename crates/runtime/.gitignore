abc-sqlite.db-wal
abc-sqlite.db-shm
