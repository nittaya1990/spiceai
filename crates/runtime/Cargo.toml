[package]
description = "The spiced runtime"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "runtime"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
app = { path = "../app" }
arrow.workspace = true
arrow-csv = "53.3.0"
arrow-flight = { workspace = true, features = ["flight-sql-experimental"] }
arrow-ipc.workspace = true
arrow-json.workspace = true
arrow-schema = { workspace = true, features = ["serde"] }
arrow_sql_gen = { path = "../arrow_sql_gen" }
arrow_tools = { path = "../arrow_tools" }
async-openai.workspace = true
async-stream.workspace = true
async-trait.workspace = true
aws-config = { version = "1.1.10", optional = true }
aws-sdk-dynamodb = { version = "1", optional = true }
aws-sdk-secretsmanager = { version = "1", optional = true }
aws-sdk-sts = { version = "1", optional = true }
axum.workspace = true
axum-extra = { version = "0.9.6", features = ["typed-header"] }
base64.workspace = true
bytes.workspace = true
cache = { path = "../cache" }
chrono = { version = "0.4.38" }
clap.workspace = true
clickhouse-rs = { workspace = true, optional = true }
csv = "1.3.0"
dashmap = "6.1.0"
data_components = { path = "../data_components" }
datafusion.workspace = true
datafusion-federation = { workspace = true }
datafusion-federation-sql = { workspace = true }
datafusion-functions-json = { workspace = true }
datafusion-table-providers = { workspace = true }
db_connection_pool = { path = "../db_connection_pool" }
document_parse = { path = "../document_parse" }
dotenvy.workspace = true
duckdb = { workspace = true, features = [
    "bundled",
    "r2d2",
    "vtab",
    "vtab-arrow",
], optional = true }
flight_client = { path = "../flight_client" }
fundu = { workspace = true }
futures.workspace = true
globset.workspace = true
governor = "0.8.0"
graph-rs-sdk = { workspace = true, optional = true }
graphql-parser.workspace = true
headers-accept = "0.1.3"
http.workspace = true
http-body-util = "0.1.2"
hyper = "1.6.0"
hyper-util = { version = "0.1.6", features = ["service"] }
iceberg.workspace = true
iceberg-catalog-rest.workspace = true
imap = { workspace = true, optional = true }
indexmap.workspace = true
itertools.workspace = true
jsonwebtoken = "9.3.0"
keyring = { version = "3.6.1", features = [
    "apple-native",
    "windows-native",
    "linux-native",
], optional = true }
llms = { path = "../llms" }
logos = "0.15.0"
mcp-client = { git = "https://github.com/block/goose", package = "mcp-client", optional = true }
mcp-core = { git = "https://github.com/block/goose", package = "mcp-core", optional = true }
mediatype = "0.19.18"
model_components = { path = "../model_components" }
mysql_async = { workspace = true, optional = true }
notify = "7.0.0"
ns_lookup = { path = "../ns_lookup" }
object_store = { workspace = true, features = ["aws", "http", "azure"] }
once_cell = "1.19.0"
opentelemetry.workspace = true
opentelemetry-prometheus.workspace = true
opentelemetry-proto = { version = "0.27", features = [
    "gen-tonic-messages",
    "gen-tonic",
    "metrics",
] }
opentelemetry_sdk.workspace = true
otel-arrow = { path = "../otel-arrow" }
package = { path = "../package" }
percent-encoding.workspace = true
pin-project.workspace = true
prometheus.workspace = true
prometheus-parse = "0.2.5"
prost = { version = "0.13.1", default-features = false, features = [
    "prost-derive",
] }
rand = "0.8.5"
regex.workspace = true
reqwest.workspace = true
runtime-auth = { path = "../runtime-auth" }
rusqlite = { workspace = true, optional = true }
rustls.workspace = true
rustls-pemfile.workspace = true
schemars = "0.8.19"
secrecy.workspace = true
serde = { workspace = true, features = ["derive"] }
serde_json.workspace = true
snafu.workspace = true
snowflake-api = { workspace = true, optional = true }
spicepod = { path = "../spicepod" }
ssh2 = { workspace = true, optional = true }
suppaftp = { workspace = true, optional = true }
telemetry = { path = "../telemetry" }
tiberius = { workspace = true, optional = true }
tokio.workspace = true
tokio-rusqlite = { workspace = true, optional = true }
tokio-rustls = "0.26.0"
tokio-stream.workspace = true
tokio-util = { workspace = true, optional = true }
tonic.workspace = true
tonic-health.workspace = true
tower.workspace = true
tower-http.workspace = true
tracing.workspace = true
tracing-futures.workspace = true
tract-core = "0.21.0"
url = "2.5.4"
util = { path = "../util" }
utoipa = { version = "5", features = ["axum_extras"], optional = true }
utoipa-swagger-ui = { version = "8", features = ["axum"], optional = true }

uuid = { workspace = true, features = ["v7"] }
x509-certificate.workspace = true

[dev-dependencies]
anyhow = "1.0.86"
async-graphql = "7.0.5"
async-graphql-axum = "=7.0.13"
azure_core = "0.21.0"
azure_storage = "0.21.0"
azure_storage_blobs = "0.21.0"
bollard = "0.18.1"
dirs = "5.0.1"
flightrepl = { path = "../flightrepl" }
hf-hub = { workspace = true }
insta.workspace = true
jsonpath-rust = "0.7.3"
opentelemetry_sdk = { workspace = true, features = ["rt-tokio-current-thread"] }
paste.workspace = true
scopeguard = "1.2.0"
serde_json = { workspace = true, features = ["preserve_order"] }
spice-cloud = { path = "../spice_cloud" }
test-framework = { path = "../test-framework" }
tokio = { workspace = true, features = ["time", "test-util"] }
tracing-opentelemetry.workspace = true
tracing-subscriber.workspace = true

[features]
aws-secrets-manager = [
    "dep:aws-config",
    "dep:aws-sdk-secretsmanager",
    "dep:aws-sdk-sts",
]
clickhouse = [
    "dep:clickhouse-rs",
    "db_connection_pool/clickhouse",
    "arrow_sql_gen/clickhouse",
    "data_components/clickhouse",
]
cuda = ["llms/cuda"]
databricks = ["data_components/databricks"]
debezium = ["data_components/debezium"]
default = ["keyring-secret-store", "aws-secrets-manager", "sharepoint"]
delta_lake = ["data_components/delta_lake"]
dev = ["openapi", "otel-arrow/dev", "mcp"]
dremio = []
duckdb = ["dep:duckdb", "db_connection_pool/duckdb", "data_components/duckdb"]
dynamodb = ["dep:aws-sdk-dynamodb", "data_components/dynamodb"]
flightsql = ["data_components/flightsql"]
ftp = ["dep:suppaftp", "dep:ssh2"]
imap = ["data_components/imap", "dep:imap"]
keyring-secret-store = ["dep:keyring"]
mcp = ["dep:mcp-client", "dep:mcp-core"]
metal = ["llms/metal"]
models = ["model_components/full"]
mssql = ["dep:tiberius", "dep:tokio-util", "data_components/mssql"]
mysql = ["dep:mysql_async", "db_connection_pool/mysql", "data_components/mysql"]
odbc = ["db_connection_pool/odbc", "data_components/odbc"]
openapi = ["dep:utoipa-swagger-ui", "dep:utoipa"]
postgres = [
    "db_connection_pool/postgres",
    "data_components/postgres",
    "datafusion-table-providers/postgres",
]
postgres-write = []
sharepoint = ["data_components/sharepoint", "dep:graph-rs-sdk"]
snowflake = [
    "dep:snowflake-api",
    "db_connection_pool/snowflake",
    "data_components/snowflake",
]
spark = ["data_components/spark_connect"]
sqlite = [
    "dep:tokio-rusqlite",
    "db_connection_pool/sqlite",
    "data_components/sqlite",
    "dep:rusqlite",
]

[[bench]]
harness = false
name = "bench"

[[bench]]
harness = false
name = "vector_search"

[target.'cfg(windows)'.dependencies]
winver = "1.0.0"
