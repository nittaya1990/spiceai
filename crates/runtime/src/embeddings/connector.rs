/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
use crate::component::dataset::Dataset;
use crate::dataconnector::DataConnectorError;
use crate::model::EmbeddingModelStore;
use async_trait::async_trait;
use datafusion::datasource::TableProvider;
use itertools::Itertools;
use llms::chunking::ChunkingConfig;
use spicepod::component::embeddings::ColumnEmbeddingConfig;
use std::any::Any;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

use crate::dataconnector::DataConnector;
use crate::dataconnector::DataConnectorResult;
use crate::model::ENABLE_MODEL_SUPPORT_MESSAGE;

use super::table::EmbeddingTable;

pub struct EmbeddingConnector {
    inner_connector: Arc<dyn DataConnector>,

    embedding_models: Arc<RwLock<EmbeddingModelStore>>,
}

impl EmbeddingConnector {
    pub fn new(
        inner_connector: Arc<dyn DataConnector>,
        embedding_models: Arc<RwLock<EmbeddingModelStore>>,
    ) -> Self {
        Self {
            inner_connector,
            embedding_models,
        }
    }

    /// Wrap an existing [`TableProvider`] with a [`EmbeddingTable`] provider. If no embeddings
    /// are needed for the [`Dataset`], it is not unnecessarily nested.
    pub(crate) async fn wrap_table(
        &self,
        inner_table_provider: Arc<dyn TableProvider>,
        dataset: &Dataset,
    ) -> DataConnectorResult<Arc<dyn TableProvider>> {
        // Runtime isn't built with model support, but user specified a dataset to use embeddings.
        if !cfg!(feature = "models") {
            return Err(DataConnectorError::InvalidConfigurationNoSource {
                dataconnector: dataset.source().to_string(),
                message: format!("The dataset is configured with an embedding model, but the runtime is not built with model support.\n{ENABLE_MODEL_SUPPORT_MESSAGE}"),
                connector_component: dataset.into()
            });
        }

        // Add in embedding columns from `dataset.columns.embeddings`.
        let from_columns: Vec<ColumnEmbeddingConfig> = dataset
            .columns
            .iter()
            .flat_map(|column| {
                column.embeddings.iter().map(|e| ColumnEmbeddingConfig {
                    column: column.name.clone(),
                    model: e.model.clone(),
                    chunking: e.chunking.clone(),
                    primary_keys: e.row_ids.clone(),
                })
            })
            .collect_vec();
        let mut embeddings = dataset.embeddings.clone();
        embeddings.extend(from_columns);

        if embeddings.is_empty() {
            return Ok(inner_table_provider);
        }

        let embed_columns: HashMap<String, String, _> = embeddings
            .iter()
            .map(|e| (e.column.clone(), e.model.clone()))
            .collect::<HashMap<_, _>>();

        // Early check if embedding models are available.
        for (column, model) in &embed_columns {
            if !self.embedding_models.read().await.contains_key(model) {
                return Err(DataConnectorError::InvalidConfigurationNoSource {
                    dataconnector: "EmbeddingConnector".to_string(),
                    message: format!("The dataset is configured with an embedding model '{model}' to embed column '{column}', but the model '{model}' is not defined in Spicepod (as an 'embeddings') or failed to load.\nFor details, visit: https://spiceai.org/docs/components/embeddings"),
                    connector_component: dataset.into()
                });
            }
        }

        let embed_chunker_config: HashMap<String, ChunkingConfig> = embeddings
            .iter()
            .filter(|e| e.chunking.as_ref().is_some_and(|s| s.enabled))
            .filter_map(|e| {
                e.chunking.as_ref().map(|chunk_cfg| {
                    (
                        e.column.clone(),
                        ChunkingConfig {
                            target_chunk_size: chunk_cfg.target_chunk_size,
                            overlap_size: chunk_cfg.overlap_size,
                            trim_whitespace: chunk_cfg.trim_whitespace,
                            file_format: dataset.params.get("file_format").map(String::as_str),
                        },
                    )
                })
            })
            .collect::<HashMap<_, _>>();

        let embedding_table = EmbeddingTable::try_new(
            inner_table_provider,
            embed_columns,
            Arc::clone(&self.embedding_models),
            embed_chunker_config,
        )
        .await
        .map_err(|e| DataConnectorError::InvalidConfiguration {
            dataconnector: dataset.source().to_string(),
            message: e.to_string(),
            connector_component: dataset.into(),
            source: Box::new(e),
        })?;

        Ok(Arc::new(embedding_table) as Arc<dyn TableProvider>)
    }
}

#[async_trait]
impl DataConnector for EmbeddingConnector {
    fn as_any(&self) -> &dyn Any {
        self
    }

    async fn read_provider(
        &self,
        dataset: &Dataset,
    ) -> DataConnectorResult<Arc<dyn TableProvider>> {
        self.wrap_table(self.inner_connector.read_provider(dataset).await?, dataset)
            .await
    }

    async fn read_write_provider(
        &self,
        dataset: &Dataset,
    ) -> Option<DataConnectorResult<Arc<dyn TableProvider>>> {
        match self.inner_connector.read_write_provider(dataset).await {
            Some(Ok(inner)) => Some(self.wrap_table(inner, dataset).await),
            Some(Err(e)) => Some(Err(e)),
            None => None,
        }
    }

    async fn metadata_provider(
        &self,
        dataset: &Dataset,
    ) -> Option<DataConnectorResult<Arc<dyn TableProvider>>> {
        self.inner_connector.metadata_provider(dataset).await
    }
}
