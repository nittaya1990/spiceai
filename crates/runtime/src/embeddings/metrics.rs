/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{sync::LazyLock, time::Duration};

use async_openai::types::{CreateEmbeddingRequest, EncodingFormat};
use opentelemetry::{
    global,
    metrics::{Counter, Histogram, Meter},
    Key, KeyValue, Value,
};

static METER: LazyLock<Meter> = LazyLock::new(|| global::meter("embeddings"));

pub(crate) static EMBEDDING_REQUESTS: LazyLock<Counter<u64>> =
    LazyLock::new(|| METER.u64_counter("embeddings_requests").build());

pub(crate) static FAILURES: LazyLock<Counter<u64>> = LazyLock::new(|| {
    METER
        .u64_counter("embeddings_failures")
        .with_description("Number of embedding failures.")
        .build()
});

pub(crate) static EMBEDDING_INTERNAL_DURATION_MS: LazyLock<Histogram<f64>> = LazyLock::new(|| {
    METER
        .f64_histogram("embeddings_internal_request_duration_ms")
        .with_unit("ms")
        .with_description("The duration of running an embedding(s) internally.")
        .build()
});

pub(crate) fn simple_labels(model: &str, encoding_format: &EncodingFormat) -> Vec<KeyValue> {
    let encoding = match encoding_format {
        EncodingFormat::Float => "float",
        EncodingFormat::Base64 => "base64",
    };
    vec![
        KeyValue::new(Key::new("model"), Value::String(model.to_string().into())),
        KeyValue::new(Key::new("encoding_format"), Value::String(encoding.into())),
    ]
}

pub(crate) fn request_labels(req: &CreateEmbeddingRequest) -> Vec<KeyValue> {
    let encoding = match req.encoding_format {
        None | Some(EncodingFormat::Float) => "float",
        Some(EncodingFormat::Base64) => "base64",
    };
    let mut labels = vec![
        KeyValue::new(
            Key::new("model"),
            Value::String(req.model.to_string().into()),
        ),
        KeyValue::new(Key::new("encoding_format"), Value::String(encoding.into())),
    ];
    if let Some(ref user) = req.user {
        labels.push(KeyValue::new(
            Key::new("user"),
            Value::String(user.to_string().into()),
        ));
    };

    if let Some(ref dim) = req.dimensions {
        labels.push(KeyValue::new(
            Key::new("dimensions"),
            Value::I64(i64::from(*dim)),
        ));
    }

    labels
}

pub(crate) fn handle_metrics(duration: Duration, is_failure: bool, labels: &[KeyValue]) {
    EMBEDDING_REQUESTS.add(1, labels);

    EMBEDDING_INTERNAL_DURATION_MS.record(duration.as_secs_f64(), labels);

    if is_failure {
        FAILURES.add(1, labels);
    }
}
