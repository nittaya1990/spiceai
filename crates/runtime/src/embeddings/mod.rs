/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
pub mod common;
pub mod connector;
pub mod cosine_distance;
pub mod execution_plan;
pub mod metrics;
pub mod table;
pub mod task;
pub mod vector_search;

/// Converts string-like Arrow types into an iterator [`Option<Box<dyn Iterator<Item = Option<&str>>>>`]. If the downcast conversion
/// fails, returns `None`.
#[macro_export]
macro_rules! convert_string_arrow_to_iterator {
    ($data:expr) => {{
        None.or($data
            .as_any()
            .downcast_ref::<StringArray>()
            .map(|arr| Box::new(arr.iter()) as Box<dyn Iterator<Item = Option<&str>> + Send>))
            .or($data
                .as_any()
                .downcast_ref::<StringViewArray>()
                .map(|arr| Box::new(arr.iter()) as Box<dyn Iterator<Item = Option<&str>> + Send>))
            .or($data
                .as_any()
                .downcast_ref::<LargeStringArray>()
                .map(|arr| Box::new(arr.iter()) as Box<dyn Iterator<Item = Option<&str>> + Send>))
    }};
}
