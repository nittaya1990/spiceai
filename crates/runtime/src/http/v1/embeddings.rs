/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::sync::Arc;

use crate::model::EmbeddingModelStore;
use async_openai::types::CreateEmbeddingRequest;
#[cfg(feature = "openapi")]
use async_openai::types::CreateEmbeddingResponse;
use axum::{
    http::StatusCode,
    response::{IntoResponse, Response},
    Extension, Json,
};
use tokio::sync::RwLock;

/// Create Embeddings
///
/// Creates an embedding vector representing the input text.
///
/// Get a vector representation of a given input that can be easily consumed by machine learning models and algorithms.
#[cfg_attr(feature = "openapi", utoipa::path(
    post,
    path = "/v1/embeddings",
    operation_id = "post_embeddings",
    tag = "AI",
    request_body(
        description = "Embedding creation request parameters",
        content((
            CreateEmbeddingRequest = "application/json",
            example = json!({
                "input": "The food was delicious and the waiter...",
                "model": "text-embedding-ada-002",
                "encoding_format": "float"
            })
        ))
    ),
    responses(
        (status = 200, description = "Embedding created successfully", content((
            CreateEmbeddingResponse = "application/json",
            example = json!({
                "object": "list",
                "data": [
                    {
                        "object": "embedding",
                        "embedding": [
                            0.002_306_425_5,
                            -0.009_327_292,
                            -0.002_884_222_2
                        ],
                        "index": 0
                    }
                ],
                "model": "text-embedding-ada-002",
                "usage": {
                    "prompt_tokens": 8,
                    "total_tokens": 8
                }
            })
        ))),
        (status = 404, description = "Model not found", content((
            String = "application/json",
            example = json!({
                "error": "model not found"
            })
        ))),
        (status = 500, description = "Internal server error", content((
            String = "application/json",
            example = json!({
                "error": "Unexpected internal server error occurred"
            })
        )))
    )
))]
pub(crate) async fn post(
    Extension(embeddings): Extension<Arc<RwLock<EmbeddingModelStore>>>,
    Json(req): Json<CreateEmbeddingRequest>,
) -> Response {
    let model_id = req.model.clone().to_string();
    match embeddings.read().await.get(&model_id) {
        Some(model) => {
            let resp: Response = match model.embed_request(req).await {
                Ok(response) => Json(response).into_response(),
                Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
            };

            resp
        }
        None => (StatusCode::NOT_FOUND, "model not found").into_response(),
    }
}
