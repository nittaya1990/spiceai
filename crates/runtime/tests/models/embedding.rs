/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#![allow(clippy::expect_used)]
use super::send_embeddings_request;
use crate::{
    models::{create_api_bindings_config, normalize_embeddings_response},
    utils::{runtime_ready_check, runtime_ready_check_with_timeout},
};
use app::AppBuilder;
use async_openai::types::{CreateEmbeddingResponse, EmbeddingInput};
use core::time;
use runtime::{auth::EndpointAuth, Runtime};
use spicepod::component::embeddings::Embeddings;
use std::sync::Arc;

pub(crate) struct EmbeddingTestCase<'a> {
    pub input: EmbeddingInput,
    pub encoding_format: Option<&'static str>,
    pub user: Option<&'static str>,
    pub model_name: &'a str,
    pub dimensions: Option<u32>,
    pub test_id: &'static str,
}

/// Handle running a set of test cases against an Embeddings model.
/// This function sets up and runs the server.
pub(crate) async fn run_embedding_tests(
    models: Vec<Embeddings>,
    tests: Vec<EmbeddingTestCase<'_>>,
) -> Result<(), anyhow::Error> {
    let (_, http_base) = start_runtime_with_embedding(models, None).await?;

    for EmbeddingTestCase {
        input,
        encoding_format,
        user,
        model_name,
        dimensions,
        test_id,
    } in tests
    {
        let response = send_embeddings_request(
            http_base.as_str(),
            model_name,
            input,
            encoding_format,
            user,
            dimensions,
        )
        .await?;

        insta::assert_snapshot!(
            format!("{model_name}_{test_id}"),
            normalize_embeddings_response(response)
        );
    }
    Ok(())
}

/// Handle tests for beta embedding criteria. Does not run performance tests needed to satisfy beta criteria.
pub(crate) async fn run_beta_functionality_criteria_test(
    model: Embeddings,
    ready_timeout: time::Duration,
) -> Result<(), anyhow::Error> {
    let model_name = model.name.clone();
    let (_rt, http_base) = start_runtime_with_embedding(vec![model], Some(ready_timeout)).await?;

    let tests = vec![
        EmbeddingTestCase {
            input: EmbeddingInput::String("The food was delicious and the waiter...".to_string()),
            model_name: model_name.as_str(),
            encoding_format: Some("float"),
            user: None,
            dimensions: None,
            test_id: "alpha_float",
        },
        EmbeddingTestCase {
            input: EmbeddingInput::String("The food was delicious and the waiter...".to_string()),
            model_name: model_name.as_str(),
            encoding_format: Some("base64"),
            user: None,
            dimensions: None,
            test_id: "alpha_base64",
        },
        EmbeddingTestCase {
            input: EmbeddingInput::StringArray(vec![
                "The food was delicious".to_string(),
                "and the waiter...".to_string(),
            ]),
            encoding_format: None,
            model_name: model_name.as_str(),
            user: None,
            dimensions: Some(256),
            test_id: "alpha_string_array",
        },
        EmbeddingTestCase {
            input: EmbeddingInput::IntegerArray(vec![83, 8251, 2488, 382, 2212, 0]),
            encoding_format: None,
            model_name: model_name.as_str(),
            user: None,
            dimensions: Some(256),
            test_id: "alpha_integer_array",
        },
        EmbeddingTestCase {
            input: EmbeddingInput::ArrayOfIntegerArray(vec![
                vec![17, 1343, 362, 796, 604],
                vec![83, 8251, 2488, 382, 2212, 0],
            ]),
            encoding_format: None,
            model_name: model_name.as_str(),
            user: None,
            dimensions: Some(256),
            test_id: "alpha_array_integer_array",
        },
    ];

    for EmbeddingTestCase {
        input,
        encoding_format,
        user: _,
        model_name,
        dimensions,
        test_id,
    } in tests
    {
        let response_raw = send_embeddings_request(
            http_base.as_str(),
            model_name,
            input,
            encoding_format,
            Some(test_id),
            dimensions,
        )
        .await?;

        let response: CreateEmbeddingResponse = serde_json::from_value(response_raw.clone())
            .unwrap_or_else(|_| {
                panic!("Failed to parse response for test {test_id} for model {model_name}.")
            });

        // Beta: Check for usage
        assert!(
            response.usage.prompt_tokens > 0,
            "Prompt tokens in usage should not be empty in response for test {test_id} and model {model_name}."
        );
        assert!(
            response.usage.total_tokens > 0,
            "Total tokens in usage should not be empty in response for test {test_id} and model {model_name}."
        );

        // Beta (TODO): Check for tracing
        // Beta (TODO): Check for metrics

        // Check consistenct of response.
        insta::assert_snapshot!(
            format!("{model_name}_{test_id}"),
            normalize_embeddings_response(response_raw)
        );
    }
    Ok(())
}

async fn start_runtime_with_embedding(
    models: Vec<Embeddings>,
    ready_timeout: Option<std::time::Duration>,
) -> Result<(Arc<Runtime>, String), anyhow::Error> {
    let mut app_builder = AppBuilder::new("embedding_app");

    for m in models {
        app_builder = app_builder.with_embedding(m);
    }
    let app = app_builder.build();

    let api_config = create_api_bindings_config();
    let http_base = format!("http://{}", api_config.http_bind_address);

    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(60)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    match ready_timeout {
        Some(timeout) => runtime_ready_check_with_timeout(&rt, timeout).await,
        None => runtime_ready_check(&rt).await,
    }

    Ok((rt, http_base))
}
