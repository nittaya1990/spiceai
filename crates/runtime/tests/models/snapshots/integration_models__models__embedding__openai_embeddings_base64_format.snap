---
source: crates/runtime/tests/models/embedding.rs
expression: normalize_embeddings_response(response)
snapshot_kind: text
---
{
  "data": [
    {
      "embedding": "str_len_684",
      "index": 0,
      "object": "embedding"
    },
    {
      "embedding": "str_len_684",
      "index": 1,
      "object": "embedding"
    }
  ],
  "model": "openai_embeddings",
  "object": "list",
  "usage": {
    "prompt_tokens": 8,
    "total_tokens": 8
  }
}
